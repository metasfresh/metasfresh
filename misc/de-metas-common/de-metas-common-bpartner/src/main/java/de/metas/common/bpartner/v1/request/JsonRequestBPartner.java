/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import static de.metas.common.rest_api.v1.SwaggerDocConstants.PARENT_SYNC_ADVISE_DOC;

import de.metas.common.rest_api.common.JsonMetasfreshId;
import de.metas.common.rest_api.common.JsonExternalId;
import de.metas.common.rest_api.v1.JsonInvoiceRule;
import de.metas.common.rest_api.v1.SyncAdvise;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@ToString
@EqualsAndHashCode
@ApiModel(description = "Note that given the respective use-case, either one of both properties might be `null`, but not both at once.")
public class JsonRequestBPartner
{
	@ApiModelProperty(position = 10,  //
			dataType = "java.lang.String", //
			value = "This translates to `C_BPartner.ExternalId`. This is deprecated and will no longer be available with v2.")
	private JsonExternalId externalId;

	@ApiModelProperty(hidden = true)
	private boolean externalIdSet;

	@ApiModelProperty(position = 20,  //
			value = "This translates to `C_BPartner.Value`.")
	private String code;

	@ApiModelProperty(hidden = true)
	private boolean codeSet;

	@ApiModelProperty(position = 25,  //
			value = "If not specified but required (e.g. because a new partner is created), then `true` is assumed.")
	private Boolean active;

	@ApiModelProperty(hidden = true)
	private boolean activeSet;

	@ApiModelProperty(position = 30,  //
			value = "This translates to `C_BPartner.Name`.\n"
					+ "If this is empty, and a BPartner with the given `name` does not yet exist, then the request will fail.")
	private String name;

	@ApiModelProperty(hidden = true)
	private boolean nameSet;

	@ApiModelProperty(position = 40,  //
			value = "This translates to `C_BPartner.Name2`.")
	private String name2;

	@ApiModelProperty(hidden = true)
	private boolean name2Set;

	@ApiModelProperty(position = 50,  //
			value = "This translates to `C_BPartner.Name3`.")
	private String name3;

	@ApiModelProperty(hidden = true)
	private boolean name3Set;

	@ApiModelProperty(position = 60, //
			value = "This translates to `C_BPartner.CompanyName`.\n"
					+ "If set, the the respective `C_BPartner` record will also have `IsCompany='Y'`")
	private String companyName;

	@ApiModelProperty(hidden = true)
	private boolean companyNameSet;

	private Boolean vendor;

	@ApiModelProperty(hidden = true)
	private boolean vendorSet;

	private Boolean customer;

	@ApiModelProperty(hidden = true)
	private boolean customerSet;

	@ApiModelProperty(position = 70, //
			value = "This translates to `C_BPartner.BPartner_Parent_ID`. It's a this bpartner's central/parent company",//
			dataType = "java.lang.Integer")
	private JsonMetasfreshId parentId;

	@ApiModelProperty(hidden = true)
	private boolean parentIdSet;

	@ApiModelProperty(position = 80, //
			value = "This translates to `C_BPartner.Phone2`. It's this bpartner's central phone number")
	private String phone;

	@ApiModelProperty(hidden = true)
	private boolean phoneSet;

	@ApiModelProperty(position = 90)
	private String language;

	@ApiModelProperty(hidden = true)
	private boolean languageSet;

	@ApiModelProperty(position = 100, //
			value = "Optional; if specified, it will be used, e.g. when an order is created for this business partner.")
	private JsonInvoiceRule invoiceRule;

	@ApiModelProperty(hidden = true)
	private boolean invoiceRuleSet;

	@ApiModelProperty(position = 110)
	private String url;

	@ApiModelProperty(hidden = true)
	private boolean urlSet;

	@ApiModelProperty(position = 120)
	private String url2;

	@ApiModelProperty(hidden = true)
	private boolean url2Set;

	@ApiModelProperty(position = 130)
	private String url3;

	@ApiModelProperty(hidden = true)
	private boolean url3Set;

	@ApiModelProperty(position = 140, //
			value = "Name of the business partner's group")
	private String group;

	@ApiModelProperty(hidden = true)
	private boolean groupSet;

	@ApiModelProperty(position = 150, //
			value = "Translates to `C_BPartner.GlobalId`")
	private String globalId;

	@ApiModelProperty(hidden = true)
	private boolean globalIdset;

	@ApiModelProperty(position = 160, //
			value = "Translates to `C_BPartner.VATaxId`")
	private String vatId;

	@ApiModelProperty(hidden = true)
	private boolean vatIdSet;

	@ApiModelProperty(position = 170, // shall be last
			value = "Sync advise about this bPartner's individual properties.\n"
			+ "IfExists is ignored on this level!\n" + PARENT_SYNC_ADVISE_DOC)
	private SyncAdvise syncAdvise;

	@ApiModelProperty(hidden = true)
	private boolean syncAdviseSet;

	public void setExternalId(final JsonExternalId externalId)
	{
		this.externalId = externalId;
		this.externalIdSet = true;
	}

	public void setCode(final String code)
	{
		this.code = code;
		this.codeSet = true;
	}

	public void setActive(final Boolean active)
	{
		this.active = active;
		this.activeSet = true;
	}

	public void setName(final String name)
	{
		this.name = name;
		this.nameSet = true;
	}

	public void setName2(final String name2)
	{
		this.name2 = name2;
		this.name2Set = true;
	}

	public void setName3(final String name3)
	{
		this.name3 = name3;
		this.name3Set = true;
	}

	public void setCompanyName(final String companyName)
	{
		this.companyName = companyName;
		this.companyNameSet = true;
	}

	public void setVendor(final Boolean vendor)
	{
		this.vendor = vendor;
		this.vendorSet = true;
	}

	public void setCustomer(final Boolean customer)
	{
		this.customer = customer;
		this.customerSet = true;
	}

	public void setParentId(final JsonMetasfreshId parentId)
	{
		this.parentId = parentId;
		this.parentIdSet = true;
	}

	public void setPhone(final String phone)
	{
		this.phone = phone;
		this.phoneSet = true;
	}

	public void setLanguage(final String language)
	{
		this.language = language;
		this.languageSet = true;
	}

	public void setInvoiceRule(final JsonInvoiceRule invoiceRule)
	{
		this.invoiceRule = invoiceRule;
		this.invoiceRuleSet = true;
	}

	public void setUrl(final String url)
	{
		this.url = url;
		this.urlSet = true;
	}

	public void setUrl2(final String url2)
	{
		this.url2 = url2;
		this.url2Set = true;
	}

	public void setUrl3(final String url3)
	{
		this.url3 = url3;
		this.url3Set = true;
	}

	public void setGroup(final String group)
	{
		this.group = group;
		this.groupSet = true;
	}

	public void setGlobalId(final String globalId)
	{
		this.globalId = globalId;
		this.globalIdset = true;
	}

	public void setSyncAdvise(final SyncAdvise syncAdvise)
	{
		this.syncAdvise = syncAdvise;
		this.syncAdviseSet = true;
	}

	public void setVatId(final String vatId)
	{
		this.vatId = vatId;
		this.vatIdSet = true;
	}
}
