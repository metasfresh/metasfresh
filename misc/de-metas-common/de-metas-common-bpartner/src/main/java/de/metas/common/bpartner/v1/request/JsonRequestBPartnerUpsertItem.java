/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import static de.metas.common.rest_api.v1.SwaggerDocConstants.BPARTNER_IDENTIFIER_DOC;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

@Value
@Builder(toBuilder = true)
@ApiModel(description = "Contains an external id and the actual bpartner to insert or update. The response will contain the given external id.")
public class JsonRequestBPartnerUpsertItem
{
	@ApiModelProperty(allowEmptyValue = false, //
			position = 10,
			value = BPARTNER_IDENTIFIER_DOC) //
	@NonNull
	String bpartnerIdentifier;

	@ApiModelProperty(allowEmptyValue = false, //
			position = 20,
			value = "The business partner to upsert. Note that its `externalId` is ignored in favor of this upsertRequest's `externalId`")
	@NonNull
	JsonRequestComposite bpartnerComposite;

	@JsonCreator
	public JsonRequestBPartnerUpsertItem(
			@NonNull @JsonProperty("bpartnerIdentifier") final String bpartnerIdentifier,
			@NonNull @JsonProperty("bpartnerComposite") final JsonRequestComposite bpartnerComposite)
	{
		this.bpartnerIdentifier = bpartnerIdentifier;
		this.bpartnerComposite = bpartnerComposite;
	}
}
