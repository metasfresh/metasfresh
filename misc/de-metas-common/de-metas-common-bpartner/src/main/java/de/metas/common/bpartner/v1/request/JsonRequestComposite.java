/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.request;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.ImmutableList;
import de.metas.common.externalreference.v1.JsonSingleExternalReferenceCreateReq;
import de.metas.common.rest_api.v1.SyncAdvise;
import de.metas.common.util.CoalesceUtil;
import de.metas.common.util.EmptyUtil;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.Value;

import javax.annotation.Nullable;

import static de.metas.common.rest_api.v1.SwaggerDocConstants.READ_ONLY_SYNC_ADVISE_DOC;
import static de.metas.common.util.CoalesceUtil.coalesce;

@ApiModel(description = "A BPartner with `n` contacts and `n` locations.\n")
@Value
public class JsonRequestComposite
{
	// TODO if an org is given, then verify whether the current user has access to the given org
	@ApiModelProperty(position = 10)
	@JsonInclude(Include.NON_NULL)
	String orgCode;

	@ApiModelProperty(position = 20)
	JsonRequestBPartner bpartner;

	@ApiModelProperty(value = "The location's GLN can be used to lookup the whole bpartner; if multiple locations with GLN are provided, then only the first one is used", //
			position = 30)
	@JsonInclude(Include.NON_EMPTY)
	@JsonProperty("locations")
	@Getter(AccessLevel.PRIVATE)
	JsonRequestLocationUpsert locations;

	@ApiModelProperty(position = 40)
	@JsonInclude(Include.NON_EMPTY)
	@JsonProperty("contacts")
	@Getter(AccessLevel.PRIVATE)
	JsonRequestContactUpsert contacts;

	@ApiModelProperty(position = 50)
	@JsonInclude(Include.NON_EMPTY)
	@JsonProperty("bankAccounts")
	@Getter(AccessLevel.PRIVATE)
	JsonRequestBankAccountsUpsert bankAccounts;

	@ApiModelProperty(value = "The Id of the business partner from an external system.", position = 60)
	@JsonInclude(Include.NON_NULL)
	JsonSingleExternalReferenceCreateReq bPartnerReferenceCreateRequest;

	@ApiModelProperty(value = "Ths advise is applied to this composite's bpartner or any of its contacts\n"
			+ READ_ONLY_SYNC_ADVISE_DOC, position = 70)
	@JsonInclude(Include.NON_NULL)
	SyncAdvise syncAdvise;

	@Builder(toBuilder = true)
	@JsonCreator
	private JsonRequestComposite(
			@JsonProperty("orgCode") @Nullable final String orgCode,
			@JsonProperty("bpartner") @Nullable final JsonRequestBPartner bpartner,
			@JsonProperty("locations") @Nullable final JsonRequestLocationUpsert locations,
			@JsonProperty("contacts") @Nullable final JsonRequestContactUpsert contacts,
			@JsonProperty("bankAccounts") @Nullable final JsonRequestBankAccountsUpsert bankAccounts,
			@JsonProperty("bpartnerReferenceCreateRequest") @Nullable final JsonSingleExternalReferenceCreateReq bPartnerReferenceCreateRequest,
			@JsonProperty("syncAdvise") final SyncAdvise syncAdvise)
	{
		this.orgCode = orgCode;
		this.bpartner = bpartner;
		this.locations = locations;
		this.contacts = contacts;
		this.bankAccounts = bankAccounts;
		this.bPartnerReferenceCreateRequest = bPartnerReferenceCreateRequest;
		this.syncAdvise = syncAdvise;
	}

	public ImmutableList<String> extractLocationGlns()
	{
		return this.locations.getRequestItems().stream()
				.map(JsonRequestLocationUpsertItem::getLocation)
				.map(JsonRequestLocation::getGln)
				.filter(gln -> !EmptyUtil.isEmpty(gln, true))
				.collect(ImmutableList.toImmutableList());
	}

	@JsonIgnore
	public JsonRequestLocationUpsert getLocationsNotNull()
	{
		return coalesce(locations, JsonRequestLocationUpsert.builder().build());
	}

	@JsonIgnore
	public JsonRequestContactUpsert getContactsNotNull()
	{
		return coalesce(contacts, JsonRequestContactUpsert.builder().build());
	}

	@JsonIgnore
	public JsonRequestBankAccountsUpsert getBankAccountsNotNull()
	{
		return CoalesceUtil.coalesce(bankAccounts, JsonRequestBankAccountsUpsert.NONE);
	}
}
