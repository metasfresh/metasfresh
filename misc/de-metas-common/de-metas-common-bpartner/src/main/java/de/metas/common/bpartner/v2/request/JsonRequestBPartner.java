/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v2.request;

import de.metas.common.bpartner.v2.common.JsonDeliveryRule;
import de.metas.common.bpartner.v2.common.JsonDeliveryViaRule;
import de.metas.common.bpartner.v2.common.JsonPaymentRule;
import de.metas.common.rest_api.common.JsonMetasfreshId;
import de.metas.common.rest_api.v2.JsonInvoiceRule;
import de.metas.common.rest_api.v2.SyncAdvise;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

import javax.annotation.Nullable;

import static de.metas.common.rest_api.v2.SwaggerDocConstants.BPARTNER_IDENTIFIER_DOC;
import static de.metas.common.rest_api.v2.SwaggerDocConstants.BPARTNER_VALUE_DOC;
import static de.metas.common.rest_api.v2.SwaggerDocConstants.PARENT_SYNC_ADVISE_DOC;
import static de.metas.common.rest_api.v2.SwaggerDocConstants.PAYMENT_TERM_IDENTIFIER_DOC;

@Getter
@ToString
@EqualsAndHashCode
@Schema(description = "Note that given the respective use-case, either one of both properties might be `null`, but not both at once.")
public class JsonRequestBPartner
{
	@Schema(description = BPARTNER_VALUE_DOC)
	private String code;

	@Schema(hidden = true)
	private boolean codeSet;

	@Schema(description = "If not specified but required (e.g. because a new partner is created), then `true` is assumed.")
	private Boolean active;

	@Schema(hidden = true)
	private boolean activeSet;

	@Schema(description = "This translates to `C_BPartner.Name`.\n"
					+ "If this is empty, and a BPartner with the given `name` does not yet exist, then the request will fail.")
	private String name;

	@Schema(hidden = true)
	private boolean nameSet;

	@Schema(description = "This translates to `C_BPartner.Name2`.")
	private String name2;

	@Schema(hidden = true)
	private boolean name2Set;

	@Schema(description = "This translates to `C_BPartner.Name3`.")
	private String name3;

	@Schema(hidden = true)
	private boolean name3Set;

	@Schema(description = "This translates to `C_BPartner.CompanyName`.\n"
					+ "If set, the the respective `C_BPartner` record will also have `IsCompany='Y'`")
	private String companyName;

	@Schema(hidden = true)
	private boolean companyNameSet;

	private Boolean vendor;

	@Schema(hidden = true)
	private boolean vendorSet;

	private Boolean customer;

	@Schema(hidden = true)
	private boolean customerSet;

	@Schema(description = "This translates to `C_BPartner.BPartner_Parent_ID`. It's a this bpartner's central/parent company")
	private JsonMetasfreshId parentId;

	@Schema(hidden = true)
	private boolean parentIdSet;

	@Schema(description = "This translates to `C_BPartner.Phone2`. It's this bpartner's central phone number")
	private String phone;

	@Schema(hidden = true)
	private boolean phoneSet;

	@Schema
	private String language;

	@Schema(hidden = true)
	private boolean languageSet;

	@Schema(description = "Optional; if specified, it will be used, e.g. when an order is created for this business partner.")
	private JsonInvoiceRule invoiceRule;

	@Schema(hidden = true)
	private boolean invoiceRuleSet;

	@Schema(description = "Optional; if specified, it will be used, e.g. when a purchase order is created for this business partner.")
	private JsonInvoiceRule poInvoiceRule;

	@Schema(hidden = true)
	private boolean poInvoiceRuleSet;

	@Schema
	private String url;

	@Schema(hidden = true)
	private boolean urlSet;

	@Schema
	private String url2;

	@Schema(hidden = true)
	private boolean url2Set;

	@Schema
	private String url3;

	@Schema(hidden = true)
	private boolean url3Set;

	@Schema(description = "Name of the business partner's group")
	private String group;

	@Schema(hidden = true)
	private boolean groupSet;

	@Schema(description = "Translates to `C_BPartner.GlobalId`")
	private String globalId;

	@Schema(hidden = true)
	private boolean globalIdset;

	@Schema(description = "The ID of the price list that the business partner is to use as a customer. If provided, this will be used to lookup `C_BPartner.M_PricingSystem_ID`")
	private JsonMetasfreshId priceListId;

	private boolean priceListIdSet;

	@Schema(description = "Translates to `C_BPartner.VATaxId`")
	private String vatId;

	@Schema(hidden = true)
	private boolean vatIdSet;

	@Schema(description = "This translates to `C_BPartner.memo`")
	private String memo;

	private boolean memoIsSet;

	@Schema(description = "Translates to `M_SectionCode.Value` and it's mapped to `C_BPartner.M_SectionCode_ID`")
	private String sectionCodeValue;

	private boolean sectionCodeValueSet;

	@Schema(description = "Translates to `C_BPartner.Description`")
	private String description;

	private boolean descriptionSet;

	@Schema(description = "Translates to `C_BPartner.DeliveryRule`")
	private JsonDeliveryRule deliveryRule;

	private boolean deliveryRuleSet;

	@Schema(description = "Translates to `C_BPartner.DeliveryViaRule`")
	private JsonDeliveryViaRule deliveryViaRule;

	private boolean deliveryViaRuleSet;

	@Schema(description = "Translates to `C_BPartner.IsStorageWarehouse`")
	private Boolean storageWarehouse;

	private boolean storageWarehouseSet;

	@Schema(description = "Translates to `C_Incoterms.Value` and it's mapped to `C_BPartner.C_Incoterms_Customer_ID`")
	private String incotermsCustomerValue;

	private boolean incotermsCustomerValueSet;

	@Schema(description = "Translates to `C_Incoterms.Value` and it's mapped to `C_BPartner.C_Incoterms_Vendor_ID`")
	private String incotermsVendorValue;

	private boolean incotermsVendorValueSet;

	@Schema(description = "Mapped to `C_BPartner.C_PaymentTerm_ID`. " + PAYMENT_TERM_IDENTIFIER_DOC)
	private String customerPaymentTermIdentifier;

	private boolean customerPaymentTermIdentifierSet;

	@Schema(description = "Mapped to `C_BPartner.PO_PaymentTerm_ID`. " + PAYMENT_TERM_IDENTIFIER_DOC)
	private String vendorPaymentTermIdentifier;

	private boolean vendorPaymentTermIdentifierSet;

	@Schema(description = "Mapped to `C_BPartner.BPartner_Parent_ID`. " + BPARTNER_IDENTIFIER_DOC)
	private String parentIdentifier;

	private boolean parentIdentifierSet;

	@Schema(description = "Translates to `C_BPartner.PaymentRule`")
	private JsonPaymentRule paymentRule;

	private boolean paymentRuleSet;

	@Schema(description = "Translates to `C_BPartner.PaymentRulePO`")
	private JsonPaymentRule paymentRulePO;

	private boolean paymentRulePOSet;

	@Schema(description = "Mapped to `C_BPartner.Section_Group_Partner_ID`. " + BPARTNER_IDENTIFIER_DOC)
	private String sectionGroupPartnerIdentifier;

	@Schema(hidden = true)
	private boolean sectionGroupPartnerIdentifierSet;

	@Schema(description = "Translates to `C_BPartner.IsProspect`")
	private Boolean prospect;

	private boolean prospectSet;

	@Schema(description = "This translates to `C_BPartner.SAP_BPartnerCode`.")
	private String sapBPartnerCode;

	@Schema(hidden = true)
	private boolean sapBPartnerCodeSet;

	@Schema(description = "This translates to `C_BPartner.IsSectionGroupPartner`.")
	private boolean sectionGroupPartner;

	@Schema(hidden = true)
	private boolean sectionGroupPartnerSet;

	@Schema(description = "This translates to `C_BPartner.IsSectionPartner`.")
	private boolean sectionPartner;

	@Schema(hidden = true)
	private boolean sectionPartnerSet;

	@Schema(description = "This translates to `C_BPartner.Fresh_Urproduzent`.")
	private boolean urproduzent;

	@Schema(hidden = true)
	private boolean urproduzentSet;

	@Schema(description = "Sync advise about this bPartner's individual properties.\n"
					+ "IfExists is ignored on this level!\n" + PARENT_SYNC_ADVISE_DOC)
	private SyncAdvise syncAdvise;

	@Schema(hidden = true)
	private boolean syncAdviseSet;

	public void setCode(final String code)
	{
		this.code = code;
		this.codeSet = true;
	}

	public void setActive(final Boolean active)
	{
		this.active = active;
		this.activeSet = true;
	}

	public void setName(final String name)
	{
		this.name = name;
		this.nameSet = true;
	}

	public void setName2(final String name2)
	{
		this.name2 = name2;
		this.name2Set = true;
	}

	public void setName3(final String name3)
	{
		this.name3 = name3;
		this.name3Set = true;
	}

	public void setCompanyName(final String companyName)
	{
		this.companyName = companyName;
		this.companyNameSet = true;
	}

	public void setVendor(final Boolean vendor)
	{
		this.vendor = vendor;
		this.vendorSet = true;
	}

	public void setCustomer(final Boolean customer)
	{
		this.customer = customer;
		this.customerSet = true;
	}

	public void setParentId(final JsonMetasfreshId parentId)
	{
		this.parentId = parentId;
		this.parentIdSet = true;
	}

	public void setPhone(final String phone)
	{
		this.phone = phone;
		this.phoneSet = true;
	}

	public void setLanguage(final String language)
	{
		this.language = language;
		this.languageSet = true;
	}

	public void setInvoiceRule(final JsonInvoiceRule invoiceRule)
	{
		this.invoiceRule = invoiceRule;
		this.invoiceRuleSet = true;
	}

	public void setPOInvoiceRule(final JsonInvoiceRule invoiceRule)
	{
		this.poInvoiceRule = invoiceRule;
		this.poInvoiceRuleSet = true;
	}

	public void setUrl(final String url)
	{
		this.url = url;
		this.urlSet = true;
	}

	public void setUrl2(final String url2)
	{
		this.url2 = url2;
		this.url2Set = true;
	}

	public void setUrl3(final String url3)
	{
		this.url3 = url3;
		this.url3Set = true;
	}

	public void setGroup(final String group)
	{
		this.group = group;
		this.groupSet = true;
	}

	public void setGlobalId(final String globalId)
	{
		this.globalId = globalId;
		this.globalIdset = true;
	}

	public void setSyncAdvise(final SyncAdvise syncAdvise)
	{
		this.syncAdvise = syncAdvise;
		this.syncAdviseSet = true;
	}

	public void setVatId(final String vatId)
	{
		this.vatId = vatId;
		this.vatIdSet = true;
	}

	public void setMemo(final String memo)
	{
		this.memo = memo;
		this.memoIsSet = true;
	}

	public void setPriceListId(@Nullable final JsonMetasfreshId priceListId)
	{
		if (JsonMetasfreshId.toValue(priceListId) != null)
		{
			this.priceListId = priceListId;
			this.priceListIdSet = true;
		}
	}

	public void setSectionCodeValue(final String sectionCodeValue)
	{
		this.sectionCodeValue = sectionCodeValue;
		this.sectionCodeValueSet = true;
	}

	public void setDescription(final String description)
	{
		this.description = description;
		this.descriptionSet = true;
	}

	public void setDeliveryRule(final JsonDeliveryRule deliveryRule)
	{
		this.deliveryRule = deliveryRule;
		this.deliveryRuleSet = true;
	}

	public void setDeliveryViaRule(final JsonDeliveryViaRule deliveryViaRule)
	{
		this.deliveryViaRule = deliveryViaRule;
		this.deliveryViaRuleSet = true;
	}

	public void setStorageWarehouse(final Boolean storageWarehouse)
	{
		this.storageWarehouse = storageWarehouse;
		this.storageWarehouseSet = true;
	}

	public void setIncotermsCustomerValue(final String incotermsCustomerValue)
	{
		this.incotermsCustomerValue = incotermsCustomerValue;
		this.incotermsCustomerValueSet = true;
	}

	public void setIncotermsVendorValue(final String incotermsVendorValue)
	{
		this.incotermsVendorValue = incotermsVendorValue;
		this.incotermsVendorValueSet = true;
	}

	public void setCustomerPaymentTermIdentifier(final String customerPaymentTermIdentifier)
	{
		this.customerPaymentTermIdentifier = customerPaymentTermIdentifier;
		this.customerPaymentTermIdentifierSet = true;
	}

	public void setVendorPaymentTermIdentifier(final String vendorPaymentTermIdentifier)
	{
		this.vendorPaymentTermIdentifier = vendorPaymentTermIdentifier;
		this.vendorPaymentTermIdentifierSet = true;
	}

	public void setParentIdentifier(final String parentIdentifier)
	{
		this.parentIdentifier = parentIdentifier;
		this.parentIdentifierSet = true;
	}

	public void setPaymentRule(final JsonPaymentRule paymentRule)
	{
		this.paymentRule = paymentRule;
		this.paymentRuleSet = true;
	}

	public void setPaymentRulePO(final JsonPaymentRule paymentRulePO)
	{
		this.paymentRulePO = paymentRulePO;
		this.paymentRulePOSet = true;
	}

	public void setSectionGroupPartnerIdentifier(final String sectionGroupPartnerIdentifier)
	{
		this.sectionGroupPartnerIdentifier = sectionGroupPartnerIdentifier;
		this.sectionGroupPartnerIdentifierSet = true;
	}

	public void setProspect(final Boolean prospect)
	{
		this.prospect = prospect;
		this.prospectSet = true;
	}

	public void setSapBPartnerCode(final String sapBPartnerCode)
	{
		this.sapBPartnerCode = sapBPartnerCode;
		this.sapBPartnerCodeSet = true;
	}

	public void setSectionGroupPartner(final boolean sectionGroupPartner)
	{
		this.sectionGroupPartner = sectionGroupPartner;
		this.sectionGroupPartnerSet = true;
	}

	public void setSectionPartner(final boolean sectionPartner)
	{
		this.sectionPartner = sectionPartner;
		this.sectionPartnerSet = true;
	}

	public void setUrproduzent(final boolean urproduzent)
	{
		this.urproduzent = urproduzent;
		this.urproduzentSet = true;
	}
}
