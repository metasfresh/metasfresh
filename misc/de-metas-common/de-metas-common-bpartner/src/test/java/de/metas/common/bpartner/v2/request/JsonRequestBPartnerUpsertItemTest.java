/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v2.request;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.fasterxml.jackson.databind.ObjectMapper;
import de.metas.common.bpartner.v1.request.JsonRequestBPartner;
import de.metas.common.bpartner.v1.request.JsonRequestBPartnerUpsertItem;
import de.metas.common.bpartner.v1.request.JsonRequestComposite;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.io.IOException;

import static org.assertj.core.api.Assertions.*;

@ExtendWith(SnapshotExtension.class)
public class JsonRequestBPartnerUpsertItemTest
{
	private final ObjectMapper mapper = new ObjectMapper();
	private Expect expect;


	@Test
	public void serializeDeserialize() throws IOException
	{
		final JsonRequestBPartner partner = new JsonRequestBPartner();
		partner.setCode("code");

		final JsonRequestBPartnerUpsertItem item = JsonRequestBPartnerUpsertItem.builder()
				.bpartnerIdentifier("ext-12345")
				.bpartnerComposite(JsonRequestComposite.builder()
						.bpartner(partner)
						.build())
				.build();

		//final JSONObjectMapper<JsonRequestBPartnerUpsertItem> mapper = JSONObjectMapper.forClass(JsonRequestBPartnerUpsertItem.class);

		final String string = mapper.writeValueAsString(item);
		assertThat(string).isNotEmpty();

		final JsonRequestBPartnerUpsertItem result = mapper.readValue(string, JsonRequestBPartnerUpsertItem.class);

		assertThat(result).isEqualTo(item);
		expect.serializer("orderedJson").toMatchSnapshot(result);
	}
}
