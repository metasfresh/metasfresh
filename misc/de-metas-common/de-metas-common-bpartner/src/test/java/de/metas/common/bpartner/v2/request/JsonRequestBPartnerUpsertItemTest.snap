de.metas.common.bpartner.v2.request.JsonRequestBPartnerUpsertItemTest.serializeDeserialize=[
  {
    "bpartnerIdentifier" : "ext-12345",
    "bpartnerComposite" : {
      "bpartner" : {
        "externalIdSet" : false,
        "code" : "code",
        "codeSet" : true,
        "activeSet" : false,
        "nameSet" : false,
        "name2Set" : false,
        "name3Set" : false,
        "companyNameSet" : false,
        "vendorSet" : false,
        "customerSet" : false,
        "parentIdSet" : false,
        "phoneSet" : false,
        "languageSet" : false,
        "invoiceRuleSet" : false,
        "urlSet" : false,
        "url2Set" : false,
        "url3Set" : false,
        "groupSet" : false,
        "globalIdset" : false,
        "vatIdSet" : false,
        "syncAdviseSet" : false
      }
    }
  }
]